---
title: "Example code for clubrootDisplay library"
output: html_document
date: "2023-03-31"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up a location for your files

```{R}
#This is your working directory

#If you are displaying bam files (which need to be split by chromosome)
#then put their location here. The files are big, so you may want to store them locally
#The originals are in Z:/rolfe_group1/Shared/Pb_illumina/bam_files

SRA_Location<-"D:/Pb_illumina/bam_files"
```
Load the libraries
I have placed clubrootDisplay on github
You can install using devtools.
You only need to do this once

```{R}
# library("devtools")
# install_github("steve-rolfe/clubrootDisplay")
packageVersion("clubrootDisplay")
```

These libraries are needed all the time

```{R}
require("GenomicFeatures")
require("Gviz")
library(Biostrings)
library(tidyverse)
library(clubrootDisplay)
```
The package uses gff files and a chromosome to create a set of databases.
You need to create these the first time you run the package, but afterwards you can simply load them

```{R}
#Load the gff files - if you give it a location it will use your gff files - otherwise it uses the ones stored with the package
gff_files<-load_gff()
#and the genome
my_genome<-load_clubroot_genome()

#If the databases haven't been created, this creates them in the directory sql_database
sql_database<-create_sql_database(gff_files,my_genome,"./sql_database")

#load previosuly created databases
sql_database<-load_sql_database("./sql_database")
```
If, for any reason, you don't want to display a set of gene models,
either remove the corresponding database or remove it from this list

i.e. with
sql_database$PbPT3<-NULL

Display a single gene 
If you provide the models it will be quicker (so if you have a lot, then set the models once)
If you don't the function will look it up

```{R}
#We need this as the genome is not a 'standard' one
options(ucscChromosomeNames=FALSE)

my_gene<-"PLBR_LOCUS5451"

my_model<-get_model_from_gene(my_gene,sql_database)
my_model

Gviz::plotTracks(gene_GeneRegionTrack(my_gene,sql_database))
#quicker
Gviz::plotTracks(gene_GeneRegionTrack(my_gene,sql_database,my_model))
```

We can get the gene models which overlap our model and plot these

```{R}
Gviz::plotTracks(get_overlapping_genes(my_gene,sql_database))

#you can add your own tracks
my_geneTrack<-get_overlapping_genes(my_gene,sql_database) #our tracks
gTrack<-Gviz::GenomeAxisTrack() #display the genome
sTrack<-Gviz::SequenceTrack(my_genome,name = "Pbrassicae",chromosome = chromosome(my_geneTrack[[1]])) #and the sequence

#we add new tracks together
display_tracks<-c(geneTrack,gTrack,sTrack)

Gviz::plotTracks(display_tracks,showId=TRUE)

```

I have not written functions to handle the bam files as they are very big and need to be stored locally
We can handle these easily as they all have the model name in the file

```{R}
SRA_Accession<-c("PRJEB36458")
bam_files<-list.files(path=paste0(SRA_Location,"/",SRA_Accession),pattern="\\.1\\.bam$",full.names=TRUE)

#This gets the chromosome
my_chromosome<-chromosome(my_geneTrack[[1]]) #get the chromosome we want to display

#grab the appropriate bam file
bam_file_active<-bam_files[grep(my_chromosome,bam_files)]
bam_file_active

alTrack<-AlignmentsTrack(bam_file_active,isPaired = TRUE,type="pileup") #sashimi and coverage also allowed

display_tracks<-c(my_geneTrack,gTrack,sTrack,alTrack)
plotTracks(display_tracks,showId=TRUE)

```
Now deal with interproscan output
Again - give a location or the functions uses internally saved files

```{R}
ipro_list<-load_interproscan()

my_gene<-"PLBR_LOCUS5451"
my_model<-get_model_from_gene(my_gene,sql_database) #not needed but speeds things up

my_gene_ipro<-map_interproscan(my_gene,ipro_list,my_genome,sql_database,my_model)
my_gene_ipro
#This is everything - from all interproscan programs

aTrack<-AnnotationTrack(
    start=my_gene_ipro$sstart,
    width=my_gene_ipro$ssend-my_gene_ipro$sstart+1,
    chromosome=my_gene_ipro$chromosome,
    strand=my_gene_ipro$strand,
    id=my_gene_ipro$InterPro_annotations,
    group=my_gene_ipro$Signature_description
  )

my_geneTrack<-gene_GeneRegionTrack(my_gene,sql_database,my_model)

display_tracks<-c(my_geneTrack,aTrack,gTrack,sTrack)#
Gviz::plotTracks(display_tracks,showId=TRUE,
           featureAnnotation="id",
           groupAnnotation="group",
           just.group="above")

#we can filter and display ipro
my_gene_ipro<-my_gene_ipro %>% filter(InterPro_annotations!="-")

aTrack<-AnnotationTrack(
    start=my_gene_ipro$sstart,
    width=my_gene_ipro$ssend-my_gene_ipro$sstart+1,
    chromosome=my_gene_ipro$chromosome,
    strand=my_gene_ipro$strand,
    id=my_gene_ipro$InterPro_annotations,
    group=my_gene_ipro$Signature_description
  )

display_tracks<-c(my_geneTrack,aTrack,gTrack,sTrack)#
Gviz::plotTracks(display_tracks,showId=TRUE,
           featureAnnotation="id",
           groupAnnotation="group",
           just.group="above")

```

The real speed comes from doing things as lists

We can create all of the gene models and save them as png files

```{R}
my_list<-("G:/My Drive/clubrootDisplay/clubrootDisplay/example_list.txt")
my_files<-scan(my_list,what="character")
my_files
my_model<-get_model_from_gene(my_files[1],sql_database) #all are the same so get the model once 
my_model

#Create a genome track
gTrack<-Gviz::GenomeAxisTrack() #display the genome

#The png starts a png file - dev.off closes it
#I'm only doing 3 files here - remove the [1:3] to do them all

for (f in my_files[1:3]){
  png_filename=paste0(f,".png")
  print(png_filename)
  png(filename=png_filename, width=800,height=600)
  geneTrack<-get_overlapping_genes(f,sql_database) #our tracks
  sTrack<-Gviz::SequenceTrack(my_genome,name = "Pbrassicae",chromosome = chromosome(geneTrack[[1]])) #and the sequence
  my_chromosome<-chromosome(geneTrack[[1]])
  alTrack=NULL
#uncomment this if you want bam files 
  # bam_file_active<-bam_files[grep(my_chromosome,bam_files)]
  # alTrack<-AlignmentsTrack(bam_file_active,isPaired = TRUE,type="pileup") #sashimi and coverage also allowed
  display_tracks<-c(geneTrack,gTrack,sTrack,alTrack)
  plotTracks(display_tracks,showId=TRUE)
  dev.off()
}
```

