---
title: "Example code for clubrootDisplay library"
output: html_document
date: "2023-03-31"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up a location for your files

```{R}
setwd("G:/Shared drives/Clubroot_Leverhulme/communal_rcode/")
require("GenomicFeatures")
require("Gviz")
library(Biostrings)
library(clubrootDisplay)

#Load the gff files
gff_files<-load_gff()
#and the genome
my_genome<-load_clubroot_genome()
#If the databases haven't been created, this creates them
#sql_database<-create_sql_database(gff_files,my_genome,"./sql_database")

#load previosuly created databases
sql_database<-load_sql_database("./sql_database")


my_gene<-"PLBR_LOCUS5451"

#using grep is better as the gff file conversions added in extract text
my_model<-get_model_from_gene(my_gene,sql_database)


Gviz::plotTracks(gene_GeneRegionTrack(my_gene,my_model,sql_database))

#or all in one go
Gviz::plotTracks(gene_GeneRegionTrack(my_gene,
                                      get_model_from_gene(my_gene,sql_database)
                                      ,sql_database))

Gviz::plotTracks(get_overlapping_genes(my_gene,sql_database))

#you can add your own tracks
geneTrack<-get_overlapping_genes(my_gene,sql_database) #our tracks
gTrack<-Gviz::GenomeAxisTrack() #display the genome
sTrack<-Gviz::SequenceTrack(my_genome,name = "Pbrassicae",chromosome = chromosome(geneTrack[[1]])) #and the sequence



#we add new tracks
display_tracks<-c(geneTrack,gTrack,sTrack)

Gviz::plotTracks(display_tracks,showId=TRUE)

plotTracks(get_overlapping_genes("PbPT3Sc00003_Am_0.106_1",sql_database),showId=TRUE)
plotTracks(get_overlapping_genes("PLBR_LOCUS4306",sql_database),showId=TRUE)
plotTracks(get_overlapping_genes("PLBR_LOCUS4264",sql_database),showId=TRUE)

#I have not written functions to handle the bam files as they are very big and need to be stored locally

#This lists the bam files split by project and chromosome
SRA_Location<-"D:/Pb_illumina/bam_files"
SRA_Accession<-c("PRJEB36458")
bam_files<-list.files(path=paste0(SRA_Location,"/",SRA_Accession),pattern="\\.1\\.bam$",full.names=TRUE)

my_chromosome<-chromosome(geneTrack[[1]]) #get the chromosome we want to display

#grab the appropriate bam file
bam_file_active<-bam_files[grep(my_chromosome,bam_files)]

alTrack<-AlignmentsTrack(bam_file_active,isPaired = TRUE,type="pileup") #sashimi and coverage also allowed

display_tracks<-c(geneTrack,gTrack,sTrack,alTrack)
plotTracks(display_tracks,showId=TRUE)

```
